pkg/kubectl/cmd/drain.go

```diff
@@ -470,31 +464,85 @@ func (o *DrainOptions) deleteOrEvictPods(pods []api.Pod) error {
 		return err
 	}
 
+	getPodFn := func(namespace, name string) (*api.Pod, error) {
+		return o.client.Core().Pods(namespace).Get(name)
+	}
+
+	if len(policyGroupVersion) > 0 {
+		return o.evictPods(pods, policyGroupVersion, getPodFn)
+	} else {
+		return o.deletePods(pods, getPodFn)
+	}
+}
+
+func (o *DrainOptions) evictPods(pods []api.Pod, policyGroupVersion string, getPodFn func(namespace, name string) (*api.Pod, error)) error {
+	doneCh := make(chan bool, len(pods))
+	errCh := make(chan error, 1)
+
 	for _, pod := range pods {
-		if len(policyGroupVersion) > 0 {
-			err = o.evictPod(pod, policyGroupVersion)
-		} else {
-			err = o.deletePod(pod)
-		}
-		if err != nil {
+		go func(pod api.Pod, doneCh chan bool, errCh chan error) {
+			var err error
+			for {
+				err = o.evictPod(pod, policyGroupVersion)
+				if err == nil {
+					break
+				} else if apierrors.IsTooManyRequests(err) {
+					time.Sleep(5 * time.Second)
+				} else {
+					errCh <- err
+					return
+				}
+			}
+			podArray := []api.Pod{pod}
+			_, err = o.waitForDelete(podArray, kubectl.Interval, o.Timeout, true, getPodFn)
+			if err == nil {
+				doneCh <- true
+			} else {
+				errCh <- err
+			}
+		}(pod, doneCh, errCh)
+	}
+
+	doneCount := 0
+	for {
+		select {
+		case err := <-errCh:
 			return err
+		case <-doneCh:
+			doneCount++
+			if doneCount == len(pods) {
+				return nil
+			}
+		case <-time.After(o.Timeout):
```

Comment:
I think you should change the meaning of o.Timeout == 0 mean infinite, so
(1) change the flag definition text for --timeout to say "zero means infinite" instead of "zero means determine a timeout from the size of the object"
(2) pass a local variable here (instead of o.Timeout) that is computed as:
if (o.Timeout == 0) then the max value an int64 can hold
else o.Timeout